[pytorch] add test/mobile/op_deps project for dependency analysis test

ljk53 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29718 [pytorch][script] op dependency analysis bash driver
* #29550 [pytorch] add LLVM code analyzer in order to replace static dispatch
* **#29716 [pytorch] add test/mobile/op_deps project for dependency analysis test**
* #29837 [pytorch][cmake] add static libraries to TorchConfig.cmake.in
* #29715 [pytorch] set USE_STATIC_DISPATCH outside cmake

Summary:
Move out the test project from PR #29550 into this separate PR.

It serves these purposes:
- Defines the ".yaml" format to describe inter-op dependency.
- Can be used as a small testbed for us to quickly experiment, evaluate
  and test different dependency analysis techniques (llvm-pass, linker,
  etc).
- Covers various different c10 operator APIs and builds a runnable binary.

I create a 'mobile' folder under 'test/' because I feel we can create a
few other similar projects here to test mobile specific yet platform
independent stuff, e.g.:
- use host tool chain + mobile build options to do continuous E2E test;
- test custom build workflow for mobile;

Test Plan:
- run build script and verify the binary is runnable:
```
scripts/build_mobile.sh
test/mobile/op_deps/build.sh
```

Differential Revision: [D18474641](https://our.internmc.facebook.com/intern/diff/D18474641)

ljk53 committed:
[pytorch] add test/mobile/op_deps project for dependency analysis test

Summary:
Move out the test project from PR #29550 into this separate PR.

It serves these purposes:
- Defines the ".yaml" format to describe inter-op dependency.
- Can be used as a small testbed for us to quickly experiment, evaluate
  and test different dependency analysis techniques (llvm-pass, linker,
  etc).
- Covers various different c10 operator APIs and builds a runnable binary.

I create a 'mobile' folder under 'test/' because I feel we can create a
few other similar projects here to test mobile specific yet platform
independent stuff, e.g.:
- use host tool chain + mobile build options to do continuous E2E test;
- test custom build workflow for mobile;

Test Plan:
- run build script and verify the binary is runnable:
```
scripts/build_mobile.sh
test/mobile/op_deps/build.sh
```

[ghstack-poisoned]

ezyang commented:
This one needs to get hooked up to CI too.

ezyang commented on test/mobile/op_deps/CMakeLists.txt:
```diff
@@ -0,0 +1,48 @@
+project(op_deps_project)
+
+# Find torch library
+add_library(torch UNKNOWN IMPORTED)
```
For my information: how come you need this line before the find_package? I would have expected find_package to produce the library on its own without extra prompting

ljk53 replied:
For whatever reason it only searches for "Caffe2" cmake package when it's building shared lib:
https://github.com/pytorch/pytorch/blob/master/cmake/TorchConfig.cmake.in#L40
And "torch" target is defined in caffe2/CMakeLists.txt:
https://github.com/pytorch/pytorch/blob/master/caffe2/CMakeLists.txt#L642

So it seems to me the static library isn't properly supported as you mentioned: #21737
I don't know how involved it might be to fix it so figured this workaround...

ezyang replied:
Looks like TorchConfig.cmake needs some updating. Can you file a bug about this? Seems worth fixing.

ljk53 replied:
Sure, I just made an attempt to fix this problem: #29837. Let's see if it can pass all CIs.

ezyang commented on test/mobile/op_deps/CMakeLists.txt:
```diff
@@ -0,0 +1,48 @@
+add_library(torch UNKNOWN IMPORTED)
+find_package(Torch REQUIRED)
+include_directories(${TORCH_INCLUDE_DIRS})
+link_directories(${TORCH_LIBRARIES})
```
If this ever turns into a real tutorial example that people will copy paste off of, it will be better to use `target_include_directories` to encourage people to write it in the modern cmake style

ezyang commented on test/mobile/op_deps/CMakeLists.txt:
```diff
@@ -0,0 +1,48 @@
+# Main executable
+add_executable(main main.cc)
+
+# Need manually link all these static libraries:
```
When I was working on #29731 I noticed a few places in your iOS tests where you are hard coding the transitive closure of libraries. This is generally an antipattern because it means whenever another dependency is added, we have to find all of the places where we've redundantly specified the transitive closure of libraries and fix them up one-by-one. cmake is supposed to be able to keep track of transitive dependencies, so that you just link against torch, and "It Just Works"; at least, our server-side cmake is working this way. It would be good to figure out why this is not working for the mobile cmake.

Also possibly related: https://github.com/pytorch/pytorch/issues/21737

ljk53 replied:
I 100% agree. I thought cmake should be able to track of transitive dependencies, but from what I read it seems during build time we manually track and maintain the dependency with variables like Caffe2_DEPENDENCY_LIBS and link them all into the final shared library (torch.so). And in the public cmake file we add the 2 shared libraries (torch.so, c10.so) manually: https://github.com/pytorch/pytorch/blob/master/cmake/TorchConfig.cmake.in#L56

Shall I try moving these static libraries to "cmake/TorchConfig.cmake.in"?
I know how to test it with mobile build mode but not so sure about non-mobile build that involves other dependent static libraries...

ezyang replied:
> during build time we manually track and maintain the dependency with variables like Caffe2_DEPENDENCY_LIBS

This is true, but that is because we are building a single mega-torch library, and cmake doesn't really understand how to link a bunch of libraries compositionally into one big shared library (I think it's hypothetically possible to build a shared library out of a bunch of static libraries but cmake sure as hell doesn't make it easy). But once you actually make a shared library, and declare that it has public dependencies on other shared libraries, cmake will do the propagation correctly.

TorchConfig.cmake is probably wrong, but I wouldn't try to fix it in this patchset.

ezyang commented on test/mobile/op_deps/CMakeLists.txt:
```diff
@@ -0,0 +1,48 @@
+  -Wl,--whole-archive
+  SimpleOps
+  ${torch}
+  -Wl,--no-whole-archive
```
Hypothetically, the `torch_library` target is supposed to handle this for you (that's what the caffe2_interface_library macro is supposed to do.)

ljk53 replied:
I see... this line is also in caffe2/CMakeLists.txt - maybe I should try removing this condition? https://github.com/pytorch/pytorch/blob/master/cmake/TorchConfig.cmake.in#L39

ezyang replied:
Not sure :>

ljk53 replied:
This is the only thing I didn't fix in the new PR I created - will try it later.

ezyang commented on test/mobile/op_deps/simple_ops.cpp:
```diff
@@ -0,0 +1,103 @@
+  .op(torch::RegisterOperators::options()
+    .schema("aten::AA(Tensor self) -> Tensor")
+    .kernel<decltype(AA_op), &AA_op>(TensorTypeId::CPUTensorId)
+    .aliasAnalysis(c10::AliasAnalysisKind::FROM_SCHEMA))
```
@smessmer don't you think these registration lines look sooooo ugly????! XD

ezyang reviewed (approved):
What's a little cmake between friends

ezyang commented:
I don't think any of the cmake comments I pointed out have to be resolved in this PR, except maybe the addition of more comments.

ljk53 committed:
Update on "[pytorch] add test/mobile/op_deps project for dependency analysis test"


Summary:
Move out the test project from PR #29550 into this separate PR.

It serves these purposes:
- Defines the ".yaml" format to describe inter-op dependency.
- Can be used as a small testbed for us to quickly experiment, evaluate
  and test different dependency analysis techniques (llvm-pass, linker,
  etc).
- Covers various different c10 operator APIs and builds a runnable binary.

I create a 'mobile' folder under 'test/' because I feel we can create a
few other similar projects here to test mobile specific yet platform
independent stuff, e.g.:
- use host tool chain + mobile build options to do continuous E2E test;
- test custom build workflow for mobile;

Test Plan:
- run build script and verify the binary is runnable:
```
scripts/build_mobile.sh
test/mobile/op_deps/build.sh
```

Differential Revision: [D18474641](https://our.internmc.facebook.com/intern/diff/D18474641)

[ghstack-poisoned]

ljk53 commented:
> I don't think any of the cmake comments I pointed out have to be resolved in this PR, except maybe the addition of more comments.

too bad... I think I fixed most of your comments before seeing this comment :p
